### PR TITLE
Feature: support the Gateway API for the canary

### PR DESCRIPTION
**Pull request by zmberg**
Signed-off-by: barnettZQG <barnett.zqg@gmail.com>
### Ⅰ. Describe what this PR does
Feature: support the Gateway API for the canary
```yaml
apiVersion: rollouts.kruise.io/v1alpha1
kind: Rollout
metadata:
  name: rollouts-demo
spec:
  objectRef:
    workloadRef:
      apiVersion: apps/v1
      kind: Deployment
      name: echoserver
  strategy:
    canary:
      steps:
      - weight: 20
        pause: {}
      trafficRoutings:
      - service: echoserver
        gateway:
          httpRouteName: echoserver
```

### CODE REVIEW

File: pkg/controller/rollout/trafficrouting/interface.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+// Controller common function across all TrafficRouting implementation
+type Controller interface {
+	// Initialize will validate the traffic routing resource
+	Initialize(ctx context.Context) (bool, error)
+	// EnsureRoutes check and set canary desired weight. desiredWeight[0,100]
+	// 1. check if canary has been set desired weight.
+	// 2. If not, set canary desired weight
+	// When desiredWeight=0, it means that rollout has been completed and the final traffic switching process is in progress,
+	// and the canary weight should be set to 0 at this time.
```
Comment (furykerry) [inline pkg/controller/rollout/trafficrouting/interface.go:29]: plz comment about the return value, especially the bool one

File: pkg/controller/rollout/trafficrouting/interface.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+*/
+
+package trafficrouting
+
+import "context"
+
+// Controller common function across all TrafficRouting implementation
+type Controller interface {
+	// Initialize will validate the traffic routing resource
+	Initialize(ctx context.Context) (bool, error)
```
Comment (furykerry) [inline pkg/controller/rollout/trafficrouting/interface.go:24]: plz comment about the return value, especially the bool one

**Review by furykerry — COMMENTED** [review]
/lgtm

**Review by FillZpp — COMMENTED** [review]
All logs should start with upper case.

File: api/v1alpha1/rollout_types.go [hunk 3]
```diff
@@ -94,15 +94,19 @@ type CanaryStrategy struct {
...
 
 // CanaryStep defines a step of a canary workload.
 type CanaryStep struct {
 	// SetWeight sets what percentage of the canary pods should receive
-	Weight int32 `json:"weight,omitempty"`
+
+	// +optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=100
+	Weight *int32 `json:"weight,omitempty"`
```
[the remark this answers is not in the document]
Comment (zmberg) [reply api/v1alpha1/rollout_types.go:109]: This field is not required, for example, for batch release scenarios

[the remark this answers is not in the document]
Comment (FillZpp) [reply api/v1alpha1/rollout_types.go:109]: It seems to be required according to the webhook change?

```golang
		if s.Weight == nil || *s.Weight < 0 || *s.Weight > 100 {
			return field.ErrorList{field.Invalid(fldPath.Index(i).Child("Weight"),
				s.Weight, `Weight must be a positive number with "0" <= weight <= "100"`)}
		}
```

[the remark this answers is not in the document]
Comment (zmberg) [reply api/v1alpha1/rollout_types.go:109]: done

File: pkg/webhook/rollout/validating/rollout_create_update_handler.go [hunk 4]
```diff
@@ -223,7 +229,7 @@ func validateRolloutSpecCanarySteps(steps []appsv1alpha1.CanaryStep, fldPath *fi
 	for i := 1; i < stepCount; i++ {
 		prev := &steps[i-1]
 		curr := &steps[i]
-		if curr.Weight < prev.Weight {
+		if *curr.Weight < *prev.Weight {
```
Comment (FillZpp) [inline pkg/webhook/rollout/validating/rollout_create_update_handler.go:232]: Can these `Weight` be nil?

**Review by FillZpp — APPROVED** [review]
/lgtm

**Pull request merged**